### PR TITLE
docs: Add parallel processing feature proposal for issue #8

### PR DESCRIPTION
**Pull request by krrish175-byte**
## What type of PR is this?

/kind documentation
/kind proposal

## What this PR does / why we need it:

This PR adds a comprehensive technical proposal for implementing parallel processing of test cases in Ianvs core, as requested in issue #8 and discussed in PR #308.

The proposal addresses the need to reduce benchmarking time when testing multiple parameter configurations or algorithms. Currently, Ianvs executes test cases serially, which can lead to excessive execution times (e.g., 5 test cases × 2 hours each = 10 hours total). This feature will enable concurrent execution across multiple CPU cores, significantly reducing total benchmarking time.

### Proposal Contents:

- **Motivation & Problem Statement**: Why parallel processing is essential for all Ianvs developers
- **Architecture Design**: Detailed technical design showing integration with Ianvs core components
- **Backward Compatibility Analysis**: Comprehensive demonstration that all existing examples continue to work without modification
- **Impact Assessment**: Analysis of how this affects current running examples across all scenarios
- **Testing & Validation Strategy**: Plan for validating all existing examples in both serial and parallel modes
- **Implementation Roadmap**: Phased 4-week approach with clear milestones
- **Risk Assessment**: Identified risks and mitigation strategies
- **Performance Analysis**: Expected speedup metrics (3-6x for typical workloads)

### Key Design Principles:

1. **Backward Compatible**: Parallel execution is opt-in; default behavior remains serial
2. **Zero Breaking Changes**: All existing examples and workflows continue to function unchanged
3. **Robust Error Handling**: Failures in one test case don't crash the entire benchmarking job
4. **User Control**: Flexible configuration via CLI flags and YAML configuration

## Which issue(s) this PR fixes:

Related to #8
Related to #308

## Special notes for reviewers:

This proposal is intended for community review and architectural discussion before proceeding with implementation, as requested by @MooreZheng in PR #308.

The proposal will be presented at the next KubeEdge SIG AI meeting for formal architectural review. Feedback on the following aspects is particularly welcome:

- Overall architecture and design approach
- Impact analysis completeness
- Testing and validation strategy
- Implementation timeline and phases
- Any concerns about backward compatibility or existing examples

## Next Steps:

1. Community review of this proposal
2. Presentation at KubeEdge SIG AI meeting (pending schedule)
3. Revisions based on community feedback
4. Approval for implementation
5. Update PR #308 with implementation based on approved design

---

**Note**: This is a documentation-only PR. The actual implementation will be submitted separately after proposal approval.

/cc @MooreZheng @hsj576 @Poorunga
/assign @MooreZheng

### CODE REVIEW

**Comment by krrish175-byte** [comment]
Hi @MooreZheng, can you please provide updates on this pr?

**Comment by MooreZheng** [comment]
> Hi @MooreZheng, can you please provide updates on this pr?

Welcome, Krrish. The work will be appreciated. 

This is related to issue https://github.com/kubeedge/ianvs/issues/8. Please note that parallel processing is an important feature in ianvs core code that will impact all ianvs developers' work, past, present, and future. For such important features, you need to provide a proposal for community reviewers to show how it would affect all current running examples. Then a formal presentation is needed to launch a review of architecture design in the KubeEdge SIG AI before getting to any implementation. 

See a proposal example in https://github.com/kubeedge/ianvs/blob/main/docs/proposals/scenarios/GovDoc2Poster/GovDoc2Poster.md

**Review by MooreZheng — CHANGES_REQUESTED** [review]
As discussed in the routine meeting, the parallel processing feature is critical for every developer of ianvs and could be the most important and challenging feature in 2026. We want to take this opportunity to thank @krrish175-byte for the hard work, comprehensive proposal, great presentation, and valuable discussion. 

I have several comments in hand

1. Code changed.
The changed code analysis could be moved from the appendix to the content. Note that all codes are appended instead of being modified.
 
2. Analyse possible worker OOM problem. 
We can provide default settings to users. But default settings also need to be researched, like by default, how many workers/ how large per worker would be the best. Too few workers makes slight improvement compared to the serial setting, while too many workers can produce extra overhead. 
  In the future, we might also need a dynamic worker setting.  
 
3. Analyse parallel processing for the AI scheme. 
Joint inference would be the most simple one, by using data partition and map reduce technique - we can just combine the test result. 
  But training is a different story - how to combing the training result (model parameters?). Parallel processing of lifelong learnig can be considered by its mult-model nature. Parallel processing of federated learning can be considered by it distributed-learning nature with local training and global aggregation (e.g., FebAvg). But incremental learning would be difficult: it needs to do parallel for one single model training - how would the parameters be divided and combined during training in this proposal 
  We might need to discuss on a case-by-case basis.

**Comment by krrish175-byte** [comment]
Thank you so much @MooreZheng for the positive feedback and constructive comments! I am thrilled that the proposal was well received in the routine meeting.

I will address each of your points:

## 1. Code Changes - Move from Appendix to Content

**Action:** I'll restructure the proposal to move the code analysis from Appendix A into the main Implementation Details section with more context and explanation.

**Updated Structure:**
- Section 7 (Implementation Details) will include:
  - Detailed code changes with before/after comparisons
  - Explanation of why each file is modified
  - Emphasis that code is being **appended** (not modified)
  - Clear separation between "unchanged code" and "new code"

I will update this in the next revision.

## 2. Worker OOM Problem & Default Settings Research

I completely agree that default settings need empirical research. Here's my proposed approach:

### Phase 1: Conservative Defaults
- Use `cpu_count - 1` as initial safe default
- Add runtime memory monitoring with warnings

### Phase 1.5: Systematic Research (Weeks 5-7)
I propose conducting systematic experiments to determine optimal defaults:

**Experiment Design:**
1. **Worker Count vs Performance**
   - Test configurations: 1, 2, 4, 6, 8, 12, 16 workers
   - Workloads: All example benchmarks (PCB-AOI, Robot, LLM, etc.)
   - Measure: Speedup, CPU utilization, memory usage, swap activity
   
2. **Memory Profiling**
   - Profile memory per test case for each example
   - Determine: `optimal_workers = floor(available_RAM * 0.8 / memory_per_test)`
   - Document memory requirements by workload type

3. **OOM Detection & Prevention**
   - Implement memory monitoring in the code
   - Add warnings when workers * estimated_memory > available_RAM
   - Provide clear error messages with recommended worker reduction

**Deliverables:**
- Research report: "Optimal Worker Configuration for Ianvs Benchmarks"
- Decision matrix: Worker recommendations by workload type
- Updated defaults based on empirical data
- Documentation: "How to Choose Worker Count" guide

**Timeline:**
- Week 5-6: Run systematic experiments
- Week 7: Analyze results and document findings
- Week 8: Update proposal and defaults based on research

### Dynamic Worker Setting (Future Enhancement)
As you suggested, we could add dynamic worker adjustment:
```python
# Monitor memory during execution
if memory_usage > threshold:
    reduce_workers()
    LOGGER.warning("Memory pressure detected, reducing workers to {new_count}")
```

**One question** Should this empirical research be completed **before** Phase 1 merge, or can we release with conservative defaults and conduct research immediately after?

## 3. Parallel Processing for AI Schemes

This is an excellent point about different learning paradigms. Let me address each:

### Joint Inference (Simple - Covered)
Its already supported
- Data partition + map-reduce works perfectly
- Each worker processes different test samples
- Results are combined (e.g., accuracy aggregation)
- **No special handling needed**

### Lifelong Learning (Multi-Model - Covered)
It is supported by multi-model nature
- Each test case trains a different model variant
- Models are independent (no parameter sharing during parallel execution)
- Results combined after training
- **Works naturally with current design**

### Federated Learning (Distributed Training - Covered)
It is supported by distributed-learning nature
- Each test case does local training
- Global aggregation (e.g., FedAvg) happens after local training completes
- The parallel processing is at the **test case level**, not within training
- **Current design supports this**

### Incremental Learning (Complex - Needs Discussion)
**This is the challenging case**

**The Problem:**
Incremental learning trains **one model progressively** across multiple rounds. Parallelizing this requires:
- Dividing model parameters during training
- Synchronizing gradients across workers
- Combining parameter updates

**Current Proposal Scope:**
This proposal parallelizes **test cases** (different algorithm configurations), not **intra-model training** (parameter parallelism).

**Example:**
-  **Covered:** Testing 5 different incremental learning algorithms in parallel (each with its own model)
-  **Not Covered:** Parallelizing the training of a single incremental learning model across multiple workers

**Discussion Needed:**

We need to discuss this case-by-case. Here are the options:

**Option 1: Keep Out of Scope for Phase 1**
- Current proposal: Parallel **test cases**, not parallel **training**
- Incremental learning with single model -> runs serially within that test case
- Users can still parallelize by testing multiple incremental learning configurations

**Option 2: Add Intra-Training Parallelism (Major Feature)**
- This would require:
  - Model parameter partitioning
  - Gradient synchronization
  - Distributed training framework integration (e.g., PyTorch DDP, Horovod)
- This is a **much larger scope** - probably a separate proposal

**Option 3: Hybrid Approach**
- Phase 1: Test case parallelism (current proposal)
- Phase 2: Research and propose intra-training parallelism for incremental learning
- Document limitations for incremental learning in Phase 1

**My Recommendation:**

For Phase 1, I suggest:
1. Focus on **test case parallelism** (different algorithm configs)
2. Document in the proposal: "Intra-model parameter parallelism for incremental learning is out of scope for Phase 1"
3. Add to Future Work: "Phase 3: Intra-Training Parallelism for Single-Model Scenarios"
4. For incremental learning benchmarks: Users can parallelize by testing multiple incremental configurations (different hyperparameters, different algorithms)

**I have some questions** 
- Should intra-training parallelism be a **requirement** for Phase 1?
- Or can we scope it as Phase 2/3 and focus Phase 1 on test case parallelism?
- Should we schedule a specific discussion session about incremental learning parallelization?

---

## Next Steps

1. **Immediate (This Week):**
   - Update proposal with code changes moved to main content
   - Add detailed section on Worker OOM prevention
   - Add section clarifying AI scheme parallelization support and limitations
   - Add empirical research plan for optimal defaults

2. **Follow-up Discussion:**
   - Schedule discussion on incremental learning parallelization approach
   - Determine if intra-training parallelism should be in Phase 1 or later

3. **Timeline:**
   - Post updated proposal: By end of this week
   - Address any additional feedback: Week 2
   - Begin implementation: Week 3 (pending approval)

Thank you again for the thorough review and excellent feedback! This discussion is making the proposal much stronger.

Looking forward to your thoughts on the questions above, particularly around:
1. Timing of empirical worker research (before or after Phase 1 merge?)
2. Scope of incremental learning support (Phase 1 or separate proposal?)

I hope to see you all in the next meeting. Thanks a lot for the time and efforts. I am excited to work on this issue :)

Best regards,
Krrish

**Comment by hsj576** [comment]
Thanks for putting together this proposal.
One key point to emphasize: any changes to the core code must not break any existing examples. Since this feature touches the core execution pipeline (e.g., testcasecontroller.py, benchmarkingjob.py), please make sure all current examples can still run correctly without any modification — both in the default serial mode and when the parallel feature is opted in.

**Review by MooreZheng — CHANGES_REQUESTED** [review]
1. As of the out-of-memory issue, the research proposal seems fine to me. The worker-count research finding will be helpful and we still need to do the research to find out default worker setting for at least one example. Then a initial research is needed to be done before merging the first example. We might also need a dynamic worker setting in the future.
2. However, the current research for parallel processing for different paradigms is not sufficient even for a proposal. The content of shows lack of knowledge of current ianvs implementation. May leave the research to future work.
- For joint inference, considering its one-model nature, then tensor partition and data partition could work even with map reduce technique.
- For lifelong learning, deeper dive into ianvs is needed. Its multi-module and multi-model nature might suit pipeline partition and model partition well.  
3. Our reviewer is not yet convinced for its impact to current examples. Need to combine proposals to one file. and the combined proposal shall need to show the code revision consideration and justificaiton for the compatibility design. Besides, when implemenation, demos for example are needed to made the impact validation convinced for all members. 
4. Renew Lay-0 architecture by referring to below examples:  
- https://github.com/kubeedge/ianvs/blob/main/docs/proposals/scenarios/GovDoc2Poster/GovDoc2Poster.md
- https://github.com/kubeedge/ianvs/blob/main/docs/proposals/scenarios/phys-scene-gen/phys_scene_gen_proposal.md
- https://github.com/kubeedge/ianvs/blob/main/docs/proposals/scenarios/PIPL-Compliant%20Cloud-Edge%20Collaborative%20Privacy-Preserving%20Prompt%20Processing%20Framework/optimizing-privacy-performance-equilibrium-cloud-edge-llm-systems_en_PR.md 
5. We also see a DCO error. It usually means that you forgot your sign off for commits. See report in https://github.com/kubeedge/ianvs/pull/308/checks?check_run_id=64329516901

**Comment by vibhor-5** [comment]
I’d like to continue work on this PR since the feature seems important and the discussion already has useful direction. I’ll review the existing feedback, address the requested changes, and update the implementation accordingly.

Please let me know if maintainers would prefer a fresh PR instead of continuing this one.

**Comment by Akshat-Tiwari69** [comment]
## Code Review: docs: Add parallel processing feature proposal

### Recommendation: Changes Requested (consistent with existing reviewer status)

This PR proposes parallel testcase processing for the PIPL-related LLM benchmark as a documentation-only addition. It adds a 1,070-line markdown proposal with four architecture diagram images under `docs/proposals/scenarios/PIPL-Compliant Cloud-Edge Collaborative Privacy-Preserving Prompt Processing Framework/parallel-processing/`.

### Module details

Documentation only — no changes to `privacy_preserving_llm.py`, `privacy_metrics.py`, or any Ianvs core. Changes requested status reflects reviewer feedback not yet addressed.

### Assessment

The parallel processing concept is relevant to production LLM inference throughput. However, adding parallelism on top of a broken evaluation pipeline (hardcoded metrics in issue #403, uninstallable dependencies, missing dataset) would only propagate the existing bugs faster. The priority should be resolving the correctness issues documented in #403, #404 first. Once the pipeline produces valid results, a parallel execution layer would be a meaningful addition.

Recommend addressing the reviewer's requested changes and resubmitting after the core execution issues are resolved.

### Reproduction

Documentation only — not applicable.

**Comment by MooreZheng** [comment]
Hey @krrish175-byte ,
Hope this message finds you well. We would like to share the information with you that an agenda topic has been formally applied, where Pragnya is planning to raise a proposal on the parallel processing project at the routine meeting this week.

@krrish175-byte might want to be interested in the discussion, and we also invite you to the meeting to join the review:
• Meeting time: 11th June 2026, Thursday at 16:30 (Beijing Time)
• Meeting ID: https://zoom.us/my/kubeedge

**Comment by krrish175-byte** [comment]
Hi @MooreZheng,
Thank you for the invitation and for keeping me in the loop regarding the parallel processing feature! I wasn't able to attend the June 11th meeting as my focus has shifted to other projects recently.

Since Pragnya has raised a new proposal and others like @vibhor-5 have shown interest in working on this, please feel free to use any parts of my initial proposal, diagrams, or research to move the feature forward. I am glad the foundational work helped kickstart this important feature for Ianvs! If there is any discussion on the parallel processing topic and it needs a new pair of eyes, I am more than happy to help!

Best of luck to the team with the implementation.
